Added new menu and action to copy bibtex key as url

boceckts opened the pull request:
Added a new sub menu item to the menu and the right click menu to quickly copy the BibTeX key as a hyperlink to the entries' url as described in #549.
In case of an empty url the BibTeX key will be copied by its own.

![rightclickmenuentry](https://cloud.githubusercontent.com/assets/15340757/18754389/c3912c32-80e8-11e6-9ccf-1678ba30fc16.PNG)
![menuentry](https://cloud.githubusercontent.com/assets/15340757/18754393/c5706b3a-80e8-11e6-86c4-294c3f2080d9.PNG)

Copying BibTeX keys as hyperlink with their url: source
![screen shot 2016-09-23 at 10 53 26](https://cloud.githubusercontent.com/assets/15340757/18780137/6a376e8c-817c-11e6-9618-3b832334f133.png)

Copying BibTeX keys as hyperlink with their url: markdown formated
![screen shot 2016-09-23 at 10 53 42](https://cloud.githubusercontent.com/assets/15340757/18780146/751b3ba8-817c-11e6-8f78-c4e9935f8474.png)

- [x] Internal quality assurance
- [x] Change in CHANGELOG.md described
- [x] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef

tschechlovdev commented on src/main/java/net/sf/jabref/gui/BasePanel.java:
```diff
@@ -950,6 +955,40 @@ private void copyKeyAndTitle() {
+            StringBuilder sb = new StringBuilder();
+
+            long copied = 0;
+            for (BibEntry entry : selectedEntries.stream().filter(BibEntry::hasCiteKey).collect(Collectors.toList())) {
```
I think it would be better if you extract the filtered list to a local variable

tschechlovdev commented on src/main/java/net/sf/jabref/gui/BasePanel.java:
```diff
@@ -950,6 +955,40 @@ private void copyKeyAndTitle() {
+            long copied = 0;
+            for (BibEntry entry : selectedEntries.stream().filter(BibEntry::hasCiteKey).collect(Collectors.toList())) {
+                String key = entry.getCiteKeyOptional().get();
+                String url = entry.getField(FieldName.URL).isPresent() ? entry.getField(FieldName.URL).get() : "";
```
Hm, which sin does it have to have an empty url? 
Wouldn't it be better if you also filter if the url is not empty?

boceckts replied:
Hm you are right there is not much sense in using an empty hyperlink but shouldn't at least the BibTeX key get copied?

tschechlovdev replied:
Yeah you're right. The current version looks good :+1:

tschechlovdev commented on src/main/java/net/sf/jabref/gui/BasePanel.java:
```diff
@@ -950,6 +955,40 @@ private void copyKeyAndTitle() {
+                copied++;
+            }
+
+            if (copied == 0) {
```
If you extract the list above, you could also check this before the for loop, by checking if the list is empty.

tschechlovdev commented on CHANGELOG.md:
```diff
@@ -42,6 +42,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - Importer for MODS format added
 - [#2012](https://github.com/JabRef/jabref/issues/2012) Implemented integrity check for `month` field: Should be an integer or normalized (BibLaTeX), Should be normalized (BibTeX)
 - [#1779](https://github.com/JabRef/jabref/issues/1779) Implemented integrity check for `bibtexkey` field: Empty bibtexkey
+- [#549](https://github.com/JabRef/jabref/issues/549) Added new shortcut to copy the BibTeX key as a hyperlink to its url to the clipboard
```
I thought that the url should be copied? Don't really get this from the description.

boceckts replied:
A hyperlink should be copied in the format <a href="url">key</a>, meaning the Bibtex key gets displayed but it's actually a hyperlink to the url of the entry.

tschechlovdev commented on src/main/java/net/sf/jabref/gui/BasePanel.java:
```diff
@@ -950,6 +955,40 @@ private void copyKeyAndTitle() {
+                return;
+            }
+
+            final StringSelection ss = new StringSelection(sb.toString());
```
please rename ss, for example to selection.

boceckts replied:
I just used this code snipped from the `copyKeyAndTitle` mthod but actually I think it can be replaced by the ClipboardManager's `setClipboardContents()` method.

boceckts commented:
@bartsch-dev I think you recently reworked the localization script so maybe you can help me here. When I add a new translation string to en property file, is there anything else to do than to execute `./gradlew update`? Because this gives me a syntax error in the python script...
```
Traceback (most recent call last):
  File "/Users/tobias/git/jabref/scripts/syncLang.py", line 423, in <module>
    update(extended=len(sys.argv) == 3 and sys.argv[2] == "--extended")
  File "/Users/tobias/git/jabref/scripts/syncLang.py", line 384, in update
    update_properties(main_property_file=get_main_menu_properties(), other_property_files=get_other_menu_properties())
  File "/Users/tobias/git/jabref/scripts/syncLang.py", line 334, in update_properties
    keys, not_fixed, fixed = fix_duplicates(lines=lines)
  File "/Users/tobias/git/jabref/scripts/syncLang.py", line 191, in fix_duplicates
    fixed.append("{key}={value}".format(key=key, value=value))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-6: ordinal not in range(128)
```

bartsch-dev commented:
The linked PR fixes that (I forgot to mark some strings as Unicode strings).

tobiasdiez commented on src/main/java/net/sf/jabref/gui/BasePanel.java:
```diff
@@ -961,6 +966,41 @@ private void copyKeyAndTitle() {
+            }
+            for (BibEntry entry : entriesWithKey) {
+                String key = entry.getCiteKeyOptional().get();
+                String url = entry.getField(FieldName.URL).isPresent() ? entry.getField(FieldName.URL).get() : "";
```
getField(URL).orElse("");

boceckts replied:
Done

boceckts commented:
I extracted the method to a new action class.

Braunch commented on src/main/java/net/sf/jabref/gui/actions/CopyBibTeXKeyAndLink.java:
```diff
@@ -0,0 +1,61 @@
+ * In case an entry doesn't have a BibTeX key it will not be copied.
+ * In case an entry doesn't have an url this will only copy the BibTeX key.
+ */
+public class CopyBibTeXKeyAndLink implements BaseAction {
```
Please add the suffix "Action" to the class name, since this is the convenient way to name the actions in the actions package.

tobiasdiez commented:
Thanks! 👍 Can be merged from my point of view.
